src/Formats/CapnProtoSerializer.cpp

```diff
@@ -0,0 +1,1224 @@
+                auto names = enum_values.getSetOfAllNames(to_lower);
+                std::unordered_set<String> capn_enum_names;
+
+                for (auto enumerant : enumerants)
+                {
+                    String name = enumerant.getProto().getName();
+                    capn_enum_names.insert(to_lower ? boost::algorithm::to_lower_copy(name) : name);
+                }
+
+                if (names != capn_enum_names)
+                    throw Exception(
+                        ErrorCodes::CAPN_PROTO_BAD_CAST,
+                        "The set of names in Enum from CapnProto schema is different from the set of names in ClickHouse Enum");
+            }
+        }
+
+        std::optional<capnp::DynamicValue::Reader> writeRow(const ColumnPtr & column, FieldBuilder *, size_t row_num) override
+        {
+            const auto * enum_data_type = assert_cast<const DataTypeEnum<EnumType> *>(data_type.get());
+            EnumType enum_value = assert_cast<const ColumnVector<EnumType> &>(*column).getElement(row_num);
+            if (enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_VALUES)
+                return capnp::DynamicValue::Reader(capnp::DynamicEnum(enum_schema, enum_value));
+
+            auto enum_name = enum_data_type->getNameForValue(enum_value);
+            for (const auto enumerant : enum_schema.getEnumerants())
+            {
+                if (compareEnumNames(String(enum_name), enumerant.getProto().getName(), enum_comparing_mode))
+                    return capnp::DynamicValue::Reader(capnp::DynamicEnum(enumerant));
+            }
+
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot convert ClickHouse Enum value to CapnProto Enum");
+        }
+
+        void readRow(IColumn & column, const capnp::DynamicValue::Reader & reader) override
+        {
+            auto enum_value = reader.as<capnp::DynamicEnum>();
+            auto enumerant = *kj::_::readMaybe(enum_value.getEnumerant());
+            auto enum_type = assert_cast<const DataTypeEnum<EnumType> *>(data_type.get());
+            DataTypePtr nested_type = std::make_shared<DataTypeNumber<EnumType>>();
+            switch (enum_comparing_mode)
+            {
+                case FormatSettings::CapnProtoEnumComparingMode::BY_VALUES:
+                {
+                    assert_cast<ColumnVector<EnumType> &>(column).insertValue(static_cast<EnumType>(enumerant.getOrdinal()));
+                    return;
+                }
+                case FormatSettings::CapnProtoEnumComparingMode::BY_NAMES:
+                {
+                    auto value = enum_type->getValue(String(enumerant.getProto().getName()));
+                    assert_cast<ColumnVector<EnumType> &>(column).insertValue(value);
+                    return;
+                }
+                case FormatSettings::CapnProtoEnumComparingMode::BY_NAMES_CASE_INSENSITIVE:
+                {
+                    /// Find the same enum name case insensitive.
```

Comment:
We can build this mapping once in `CapnProtoEnumSerializer` constructor. I will do it later.